File: client/c/Wall.cpp
// -*-c++-*-

#include "Wall.h"
#include "game.h"


Wall::Wall(_Wall* pointer)
{
    ptr = (void*) pointer;
}

int Wall::id()
{
  return ((_Wall*)ptr)->id;
}

int Wall::x()
{
  return ((_Wall*)ptr)->x;
}

int Wall::y()
{
  return ((_Wall*)ptr)->y;
}

int Wall::owner()
{
  return ((_Wall*)ptr)->owner;
}

int Wall::health()
{
  return ((_Wall*)ptr)->health;
}

int Wall::maxHealth()
{
  return ((_Wall*)ptr)->maxHealth;
}

int Wall::size()
{
  return ((_Wall*)ptr)->size;
}

int Wall::talk(std::string message)
{
  return wallTalk( (_Wall*)ptr, message.c_str());
}


std::ostream& operator<<(std::ostream& stream,Wall ob)
{
  stream << "id: " << ((_Wall*)ob.ptr)->id  <<'\n';
  stream << "x: " << ((_Wall*)ob.ptr)->x  <<'\n';
  stream << "y: " << ((_Wall*)ob.ptr)->y  <<'\n';
  stream << "owner: " << ((_Wall*)ob.ptr)->owner  <<'\n';
  stream << "health: " << ((_Wall*)ob.ptr)->health  <<'\n';
  stream << "maxHealth: " << ((_Wall*)ob.ptr)->maxHealth  <<'\n';
  stream << "size: " << ((_Wall*)ob.ptr)->size  <<'\n';
  return stream;
}

File: client/java/Wall.java
import com.sun.jna.Pointer;

///A pile of hard stuff that is in the way.
class Wall
{
  Pointer ptr;
  int ID;
  int iteration;
  public Wall(Pointer p)
  {
    ptr = p;
    ID = Client.INSTANCE.wallGetId(ptr);
    iteration = BaseAI.iteration;
  }

  boolean validify()
  {
    if(iteration == BaseAI.iteration) return true;
    for(int i = 0; i < BaseAI.walls.length; i++)
    {
      if(BaseAI.walls[i].ID == ID)
      {
        ptr = BaseAI.walls[i].ptr;
        iteration = BaseAI.iteration;
        return true;
      }
    }
    throw new ExistentialError();
  }
    
    //commands
    
  ///Sends a message to be visualized when this unit is selected
  int talk(String message)
  {
    validify();
    return Client.INSTANCE.wallTalk(ptr, message);
  }
    
    //getters
    
  ///Unique Identifier
  public int getId()
  {
    validify();
    return Client.INSTANCE.wallGetId(ptr);
  }
  ///The X position of the top left corner of this object.  X is horizontal
  public int getX()
  {
    validify();
    return Client.INSTANCE.wallGetX(ptr);
  }
  ///The Y position of the top left corner of this object.  Y is vertical
  public int getY()
  {
    validify();
    return Client.INSTANCE.wallGetY(ptr);
  }
  ///The owning player
  public int getOwner()
  {
    validify();
    return Client.INSTANCE.wallGetOwner(ptr);
  }
  ///How much health this unit currently has
  public int getHealth()
  {
    validify();
    return Client.INSTANCE.wallGetHealth(ptr);
  }
  ///The maximum amount of health this unit can ever have
  public int getMaxHealth()
  {
    validify();
    return Client.INSTANCE.wallGetMaxHealth(ptr);
  }
  ///The length of one side of this Unit
  public int getSize()
  {
    validify();
    return Client.INSTANCE.wallGetSize(ptr);
  }

}
